[FLINK-5975] Add volume support to flink-mesos

addisonj opened the pull request:
When using containerization, specifically, docker, it is useful to be
able to attach additional volumes, such as an NFS share.

This adds support for volumes to be attached via specifying a new config
values `mesos.resourcemanager.tasks.container.volumes`. This is comma
delimited string of `[host_path:]container_path[:RO|RW]`.

It is modeled after the spark mesos framework

Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +263,71 @@ public String toString() {
+		}
+		String[] specs = containerVolumes.get().split(",");
+		List<Protos.Volume> vols = new ArrayList<Protos.Volume>();
+		for (String s : Arrays.asList(specs)) {
```
You can just loop over the String[].

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +263,71 @@ public String toString() {
+					break;
+				default:
+					throw new IllegalArgumentException("volume specification is invalid");
+
```
remove empty line

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +263,71 @@ public String toString() {
+					throw new IllegalArgumentException("volume specification is invalid");
+
+			}
+
```
remove empty line

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosTaskManagerParameters.java:
```diff
@@ -115,6 +122,12 @@ public ContaineredTaskManagerParameters containeredParameters() {
 		return containeredParameters;
 	}
 
+
```
remove empty line

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +263,71 @@ public String toString() {
+					break;
+				case 3:
+					String mode = parts.get(2).toLowerCase().trim();
+					if (!mode.equals("rw") && !mode.equals("ro")) {
```
why not just use Protos.Volume.Moude.valueOf(mode) ?

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +263,71 @@ public String toString() {
+					vol.setHostPath(parts.get(0)).setContainerPath(parts.get(1));
+					break;
+				default:
+					throw new IllegalArgumentException("volume specification is invalid");
```
We should include in volume specification in the exception message.

addisonj commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -131,9 +134,7 @@ public int getPorts() {
-		public List<? extends ConstraintEvaluator> getHardConstraints() {
-			return null;
-		}
+		public List<? extends ConstraintEvaluator> getHardConstraints() { return null; }
```
oops, intellij tricking me

addisonj commented:
@zentol thanks for the review! I think I addressed all those, updated via an amend. Wasn't sure if you want changes as new commits or not. 

Also, the build passed locally for me, but I haven't seen it work via travis yet, keeps failing in flink core for what appears to be totally unrelated (guessing flaky tests?).

Let me know if there is anything else!

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +265,67 @@ public String toString() {
+		}
+		String[] specs = containerVolumes.get().split(",");
+		List<Protos.Volume> vols = new ArrayList<Protos.Volume>();
+		for (int i = 0; i < specs.length; i++) {
```
This can be simplified to ```for (String s : specs)``

addisonj replied:
scala has ruined me... I

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosTaskManagerParameters.java:
```diff
@@ -115,6 +122,11 @@ public ContaineredTaskManagerParameters containeredParameters() {
+	/**
+	 * Get the container volumes string
+	 */
+	public Option<String> containerVolumes() { return containerVolumes; }
```
`please move the return statement into a separate line.

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +265,67 @@ public String toString() {
+				case 3:
+					Protos.Volume.Mode mode = Protos.Volume.Mode.valueOf(parts.get(2).trim().toUpperCase());
+					if (mode.equals(Protos.Volume.Mode.RW)) {
+						vol.setMode(Protos.Volume.Mode.RW);
```
this block can be simplified to ```vol.setMode(mode)```.

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +265,67 @@ public String toString() {
+					vol.setContainerPath(parts.get(0));
+					break;
+				case 2:
+					String modeOrPath = parts.get(1).trim().toUpperCase();
```
similar to below, you can use ```Protos.Volume.Mode.valueOf````. wrap it in a try-catch block; if an IllegalArgumentException is thrown we'll assume that we're deling with the patm and not the mode.

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +265,67 @@ public String toString() {
+			Protos.Volume.Builder vol = Protos.Volume.newBuilder();
+			vol.setMode(Protos.Volume.Mode.RW);
+
+			List<String> parts = Arrays.asList(s.split(":"));
```
we don't need to wrap this in a list, we can just use a String[].

zentol replied:
The parsing logic is a bit brittle and inconsistent; if parts.size == 3 and the Mode is invalid an IAE is thrown. if parts.size == 2 and the mode is invalid we'll assign a bogus path.

A regex may be more appropriate.

zentol replied:
eh nvm; a messed up mode value will always cause issues. It's just that we can actually detect it if parts.size == 3.

addisonj replied:
I thought of a regex first, but after looking at the way the spark framework handles this as well as looking at the regex, it seemed uglier (especially for all paths...). It also doesn't buy us much. 

The biggest problem is that in the case of size == 2, since both a host path and RW/RO is optional, we can't know which one was intended if doesn't explicitly match a `RW` or `RO`.

For example
`/mnt/my_path:RWx`, while it might be logical to assume to that its a typo and what they wanted was a container path of `/mnt/my_path` and `RW` as a mode, it is just as valid for that to be a host path `/mnt/my_path` and a container path relative to the workdir of `RWx`.

The format for this is taken from spark framework, which in turn the format that the docker-cli  run command uses. Ideally, you might have something like a json or yaml description to define these with named keys to remove the ambiguity, however, that would add a lot of complexity, so this still seems reasonable with a slightly sharp edge for the end user

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +265,67 @@ public String toString() {
+				continue;
+			}
+			Protos.Volume.Builder vol = Protos.Volume.newBuilder();
+			vol.setMode(Protos.Volume.Mode.RW);
```
Is it required to call ```setMode``` at least once?

zentol commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +265,67 @@ public String toString() {
+	 *                         defines mount points for a container volume. If None or empty string, returns
+	 *                         an empty iterator
+	 */
+	public List<Protos.Volume> volumes(Option<String> containerVolumes) {
```
This method should be renamed t something like ```buildVolumes```, ```toVolumes``` or such.

zentol replied:
This method could be made static to prevent side-effects.

zentol commented on flink-mesos/src/test/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorkerTest.java:
```diff
@@ -0,0 +1,75 @@
+	LaunchableMesosWorker worker = new LaunchableMesosWorker(resolver, params, containerSpec, taskId);
+
+	@Test
+	public void volumes() throws Exception {
```
by convention these method names should being with "test", i.e "testVolumes()".

addisonj commented:
@zentol thanks for the quality reviews. The code reads a lot cleaner with the mode parsing being handled by the enum. My java is pretty rusty after all being in scala of late :)

I think that handles everything. Hopefully that makes sense for reasoning of not using a regex.

EronWright commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -262,9 +264,61 @@ public String toString() {
+						.setContainerPath(parts[1]);
+					break;
+				default:
+					throw new IllegalArgumentException("volume specification is invalid, given: " + s);
```
Ideally this validation would occur eagerly, i.e. in `MesosTaskManagerParameters::create`.  Flink emphasizes eager validation.

addisonj replied:
👍  will move it there

addisonj commented:
Okay, just moved things around, agree that its a better place, but was hesitant at first as we were building any other mesos objects earlier. Hopefully this still works. I think going forwad and adding support for other container options, such as additional network ports, additional URIs to download, that would be a better place to put them.

Also, I have one other small commit, it wasn't because the volume info wasn't getting attached soon enough. 

This now is working for me in my dev env, which is 1.2.0 with these patches applied

addisonj commented:
@EronWright @zentol minor bump on this... any other steps to get this on the path to being merged?
Don't want to let this hang out for too long so I forgot about it :)

EronWright commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java:
```diff
@@ -259,9 +262,11 @@ public String toString() {
 				throw new IllegalStateException("unsupported container type");
 		}
 		if(containerInfo != null) {
+			containerInfo.addAllVolumes(params.containerVolumes());
```
There's an unnecessary restriction here, that volumes may be used only if a container image is also used.   I bet you can use volumes with the Mesos containerizer without using an image.     The code in the `case MESOS:` block should be reorganized to always set `containerInfo = Protos.ContainerInfo.newBuilder()`.  This way, `containerInfo` will never be null.

EronWright commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosTaskManagerParameters.java:
```diff
@@ -162,11 +182,65 @@ public static MesosTaskManagerParameters create(Configuration flinkConfig) {
+						Protos.Volume.Mode mode = Protos.Volume.Mode.valueOf(parts[1].trim().toUpperCase());
+						vol.setMode(mode)
+								.setContainerPath(parts[0]);
+					} catch (IllegalArgumentException e) {
```
I gather that this code treats a two-part spec as `container:mode` first, then falls back to `host:container`.   Just curious, is there some precedent for that?  I'm not sure it makes sense;  for example, `/data:ro` would create an empty read-only volume, but what good is that?

addisonj replied:
totally agree, it is strange. But this is the same spec that docker CLI uses as well as the spark mesos framework... It doesn't seem ideal and is definitely somewhat of a sharp edge but seemed best to just use the same standard

addisonj commented:
@EronWright good suggestion in regards to containerInfo without an image name, I confirmed in the mesos docs that it should work that way. 

Lemme know if there is anything else!

EronWright commented:
@zentol are you able to get this across the finish line?

addisonj commented:
@tillrohrmann this was the PR I mentioned to you during flink foward, if you get a chance to look :)

tillrohrmann commented:
Changes look good to me. Will rebase the PR and then merge it.
